crates/types/src/traits/block_contents.rs

```diff
@@ -49,21 +59,33 @@ pub trait BlockPayload:
     type Metadata: Clone + Debug + Eq + Hash + Send + Sync;
 
     /// Encoded payload.
-    type Encode<'a>: 'a + Iterator<Item = u8> + AsRef<[u8]> + Send;
+    type Encode<'a>: 'a + Iterator<Item = u8> + Send
+    where
+        Self: 'a;
 
     /// Build a payload and associated metadata with the transactions.
+    ///
+    /// # Errors
+    /// If the transaction length conversion fails.
     fn from_transactions(
         transactions: impl IntoIterator<Item = Self::Transaction>,
-    ) -> (Self, Self::Metadata);
+    ) -> Result<(Self, Self::Metadata), BlockError>;
 
     /// Build a payload with the encoded transaction bytes and metadata.
-    fn from_bytes(encoded_transactions: Self::Encode<'_>, metadata: Self::Metadata) -> Self;
+    ///
+    /// `I` may be, but not necessarily is, the `Encode` type directly from `fn encode`.
+    fn from_bytes<I>(encoded_transactions: I, metadata: Self::Metadata) -> Self
+    where
+        I: Iterator<Item = u8>;
 
     /// Build the genesis payload and metadata.
     fn genesis() -> (Self, Self::Metadata);
 
     /// Encode the payload
-    fn encode(&self) -> Self::Encode<'_>;
+    ///
+    /// # Errors
+    /// If the transaction length conversion fails.
+    fn encode(&self) -> Result<Self::Encode<'_>, BlockError>;
```

Comment:
I'm kind of waffling on this, but I'm not sure this needs to return `Result`. I feel like there should be some consistency property where if `from_transactions` succeeds, `encode` will never fail, so all the error checking is done up front in `from_transactions`. I'm not sure, though, and I guess it's not so bad if we do return a `Result` here, since this is only called in the builder and doesn't introduce a lot of annoying error handling into HotShot proper

Reply:
> if from_transactions succeeds, encode will never fail

Yeah, I think property makes sense. On the other hand, if `from_transactions` fails, that means `encode` fails, which should return an error, right? So I'm not sure how we could make `encode` not return `Result`, unless we do the length check in `from_transactions` and assume `u32::try_from` won't fail in `encode`, which I think is a bit confusing as well.

Reply:
If `from_transactions` fails you have no block payload to call `encode` on. You just failed to build a block. But this way there is only one way to fail to build a block instead of two ways.

But you are correct, the way to implement this for `VIDBlockPayload` would be to do all the length checks in `from_transactions` and then panic if `u32::try_from` fails in `encode`, which is ugly and error prone. In the sequencer, our block type will actually contain the flat bytes, which we can just return directly in `encode`, so we don't have this problem.

I'm fine keeping the `Result`, though